src/main/java/com/abrus/user/validator/UserRegistrationValidatorImpl.java

```diff
@@ -0,0 +1,96 @@
+package com.abrus.user.validator;
+
+import com.abrus.user.model.UserRegistrationDto;
+import com.abrus.user.repository.UserRepository;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+@Data
+public class UserRegistrationValidatorImpl implements UserRegistrationValidator {
+
+    private final UserRepository userRepository;
+
+    @Override
+    public ValidationResult validate(UserRegistrationDto user) {
+        return new EmailValidationStep()
+                .with(new EmailDuplicationValidationStep(userRepository))
+                .with(new PasswordValidationStep())
+                .with(new PhoneValidationStep())
+                .validate(user);
+    }
+
+    @AllArgsConstructor
+    private static final class EmailValidationStep extends ValidationStep<UserRegistrationDto> {
+        @Override
+        public ValidationResult validate(UserRegistrationDto user) {
+            String email = user.email();
+            if (email == null || email.equals("")) {
+                return new ValidationResult(false, "Email is required");
+            }
+            if (!email.matches("[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}")) {
+                return new ValidationResult(false, "Email is not valid");
+            }
+            return checkNext(user);
+        }
+    }
+
+    @AllArgsConstructor
+    private static final class EmailDuplicationValidationStep extends ValidationStep<UserRegistrationDto> {
+
+        private final UserRepository userRepository;
+
+        @Override
+        public ValidationResult validate(UserRegistrationDto user) {
+            if (userRepository.getByEmail(user.email()).isPresent()) {
+                return ValidationResult.invalid(String.format("Email [%s] is already used", user.email()));
+            }
```

Comment:
In most cases, we are not using repository in validator classes. And in most cases validator classes are util classes, which means we are not storing any data, and all methods are static.
This check could be moved to UserServiceImpl

Reply:
But you may leave it as is